crypto/chacha/asm/chacha-x86_64.pl

```diff
@@ -65,7 +65,14 @@
 ( $xlate="${dir}../../perlasm/x86_64-xlate.pl" and -f $xlate) or
 die "can't locate x86_64-xlate.pl";
 
-$avx = 2;
+if (`$ENV{CC} -Wa,-v -c -o /dev/null -x assembler /dev/null 2>&1`
+               =~ /GNU assembler version ([2-9]\.[0-9]+)/) {
+       $avx = ($1>=2.19) + ($1>=2.22);
+}
+
+if (!$avx && `$ENV{CC} -v 2>&1` =~ /((?:clang|LLVM) version|.*based on LLVM) ([0-9]+\.[0-9]+)/) {
+       $avx = ($2>=3.0) + ($2>3.0);
+}
```

Comment:
The code works provided that `CC` environment variable is set (which I assumed we always do when we build). For example,
```
export CC=gcc-7 CXX=g++-7
cmake ....
make
```
will make the Perl scripts correctly detect the compiler/assembler.

Is my assumption about how we build wrong?

Reply:
The assumption is correct. I found I forgot to preserve environment variables when using sudo access(which is needed during test to pull some golang pkgs).

Reply:
Besides, let's say `CC` env var is set, after these `if` logics, the `avx` can be `empty""`, `1` or `2`. The current code uses hardcoded value `2` for `avx`. Bssl might have removed some assembly code blocks when `avx` == 1 or empty.

Reply:
the `avx` flag determines the highest optimization level to be compiled. For example, if `avx=2` the generated assembly will contain: generic implementation (corresponding to `avx=0`), AVX implementation (corresponding to `avx=1`), and AVX2 implementation (corresponding to `avx=2`). The choice which of the three implementations to run is done in runtime based on CPU capabilities.

BoringSSL hardcodes `avx=2` because they think there is no reason not to _generate_ the AVX2 code (will it be run or not is a runtime decision). This thinking is right, provided that your assembler is modern enough to know about AVX2. This is not the case for us, and that's why we, unfortunately, have to check if the assembler supports AVX2 before generating the code.

Reply:
the latest changes remove the automatic detection of assembler version and use the compile flag instead.